enable ethernet support on the stm32mp135f-dk

arnopo opened the pull request:
This Pull reuest enable the support of the ethernet 1 instance

Hardware details:
- the PHY is a LAN8742A with an Hardware reset based on a GPIO.
- the clock reference is an externl crystal running at 25 Mhz
- 16K of non cacheable memory in the  SYSRAM is used to store ethernet DMA Buffers and descriptors 
 
 Samples tested:
 - samples/net/ipv4_autoconf
 - samples/net/dhcpv4_client
 
 Known current limitation:
 - Ethernet2 instance not implemented 
 - sample/net/zperf not build
 
 Depends on:

- https://github.com/zephyrproject-rtos/hal_stm32/pull/320
- https://github.com/zephyrproject-rtos/zephyr/pull/96575

arnopo committed:
dts: bindings: ethernet: Add the stm32mp13 bindings

Introduce stm32mp13 bindings to support the ethernet peripherals.
The "memory-regions" property is used to reference non cacheable memory
for the ETH DMA.
The "st,ext-phyclk" property aligned Linux binding is used to specify
the PHY clock for RMII.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

arnopo committed:
dts: arm :st: Add ethernet 0 node in stm32mp13

Add the Ethernet 0 MAC and MDIO nodes in the device tree.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Signed-off-by: Arif Balik <arifbalik@outlook.com>

arnopo committed:
boards: st: stm32mp135f_dk: add ethernet nodes

The STM32MP135F-DK board embeds 2 ethernets. Enable the ethernet1
in RMII mode with associated PHY.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Signed-off-by: Arif Balik <arifbalik@outlook.com>

arnopo committed:
drivers: ethernet: stm32: add support of the stm32mp13

Add the support of the stm32mp13 ethernet:
- Adapt to HAL API
- Add support of a memory region in DT for descriptor and buffers
- Add support of PHY with/without crystal 50MHz

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Signed-off-by: Arif Balik <arifbalik@outlook.com>

erwango commented on boards/st/stm32mp135f_dk/stm32mp135f_dk_defconfig:
```diff
@@ -14,3 +14,6 @@ CONFIG_CONSOLE=y
 # UART console (overrides remote proc console)
 CONFIG_UART_CONSOLE=y
+
+# Zephyr Kernel Configuration
+CONFIG_KERNEL_DIRECT_MAP=y
```
Should this really be part of the board `_defconfig`?
Doesn't it apply to the series in general ?

arnopo replied:
Right, as we  use ` MMU_REGION_FLAT_ENTRY` in  https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/st/stm32/stm32mp13x/soc.c we need is at Soc level

maass-hamburg commented on boards/st/stm32mp135f_dk/stm32mp135f_dk_defconfig:
```diff
@@ -15,5 +15,8 @@ CONFIG_CONSOLE=y
 # UART console (overrides remote proc console)
 CONFIG_UART_CONSOLE=y
 
+# Enable I2C support
```
GPIO_MCP230XX should select I2C, instead of depending on it. see https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations

arnopo replied:
@maass-hamburg 
Seems we are facing "egg or chicken" dilemma between I2C and GPIO_MCP230XX should I replace `depends on I2C`

```diff
menuconfig GPIO_MCP230XX
	bool "MCP230XX I2C-based GPIO chip"
	default y
	depends on DT_HAS_MICROCHIP_MCP23008_ENABLED || \
	           DT_HAS_MICROCHIP_MCP23009_ENABLED || \
	           DT_HAS_MICROCHIP_MCP23016_ENABLED || \
	           DT_HAS_MICROCHIP_MCP23017_ENABLED || \
	           DT_HAS_MICROCHIP_MCP23018_ENABLED
-	depends on I2C
+	select I2C
	select GPIO_MCP23XXX
	help
	  Enable driver for MCP230XX I2C-based GPIO chip.
```

arnopo committed:
drivers: ethernet: add initial support of the LAN8742 PHY

Add basic support of the LAN8742 RMII phy. The driver is inspired
from the phy_mii generic driver, with the support of a GPIO reset.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

arnopo committed:
soc: st: stm32mp13: enable direct memory map

Enable dynamic configuration of region mapping with a 1:1
virtual-to-physical address mapping.
This configuration is aligned with the static declaration of the
MMU table using MMU_REGION_FLAT_ENTRY macro.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

erwango commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -733,6 +772,10 @@ static int eth_initialize(const struct device *dev)
 		(clock_control_subsys_t)&cfg->pclken_tx);
 	ret |= clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
 		(clock_control_subsys_t)&cfg->pclken_rx);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32mp13_ethernet)
```
Can't you check `DT_INST_CLOCKS_HAS_NAME(0, mac_clk)` instead ?

erwango commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -714,13 +743,23 @@ static int eth_initialize(const struct device *dev)
 	struct eth_stm32_hal_dev_data *dev_data = dev->data;
 	const struct eth_stm32_hal_dev_cfg *cfg = dev->config;
 	ETH_HandleTypeDef *heth = &dev_data->heth;
+	uint8_t *desc_uncached_addr;
```
```suggestion
	__maybe_unused uint8_t *desc_uncached_addr;
```

arnopo committed:
drivers: ethernet: stm32: cleanup deprecated ETH_TxPacketConfig typedef

The typedef ETH_TxPacketConfigTypeDef is used consistently in the driver.
However, some instances of ETH_TxPacketConfig have not yet been replaced.
On the STM32MP13 series, the ETH_TxPacketConfig typedef is no longer
defined. This cleanup ensures compatibility and consistency.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

arnopo committed:
drivers: gpio: invert I2C config dependency for MCP23xxx

The GPIO_MCP230XX config currently depends on the I2C config. This means
that the board/SoC must enable the I2C config to support MCP23xxx devices.
Inverting this dependency allows declaring GPIO_MCP230XX directly in the
board/SoC defconfig without explicitly enabling I2C.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Suggested-by Fin Maaß <f.maass@vogl-electronic.com>

erwango commented on west.yml:
```diff
@@ -250,7 +250,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: d768bcbf9b6f75b46eff274b343a9556c0ce108d
+      revision: b6f03b6efd92da4564b38df8cb54d7dad5f37163
```
Let's drop this, as this was updated via other PRs

erwango commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -279,7 +293,11 @@ int eth_stm32_tx(const struct device *dev, struct net_pkt *pkt)
+		HAL_ETH_TxFreeCallback(heth, (uint32_t *)ctx);
+#else
 		HAL_ETH_TxFreeCallback((uint32_t *)ctx);
+#endif
```
Proposal:
To avoid redundant ifdefery in the function headers, you can define a custom arg:
```
		HAL_ETH_TxFreeCallback(STM32_ETH_ARGS(heth, (uint32_t *)ctx));
```

With:
```
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32mp13_ethernet)
#define STM32_ETH_ARGS(foo, VARGS)  foo, VARGS
#else
#define STM32_ETH_ARGS(foo, VARGS)  VARGS
#endif
```

juickar commented on arch/arm/core/mmu/arm_mmu.c:
```diff
@@ -891,6 +891,17 @@ static int __arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flag
 	switch (flags & K_MEM_CACHE_MASK) {
 
 	case K_MEM_CACHE_NONE:
+		switch (flags & K_MEM_ARM_NC_TYPE_MASK) {
```
title of the commit is also after the signature:
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>arch: arm: mmu: allow selecting memory type for non-cacheable memory

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -36,6 +37,11 @@ extern const struct device *eth_stm32_phy_dev;
 #define __eth_stm32_desc __attribute__((section(".eth_stm32_desc")))
 #define __eth_stm32_buf  __attribute__((section(".eth_stm32_buf")))
+#elif defined(CONFIG_SOC_SERIES_STM32MP13X)
+#define ETH_DMA_REGION  DT_PHANDLE(DT_NODELABEL(mac), memory_regions)
```
access by Instance and not by label

arnopo commented on include/zephyr/arch/arm/arm_mem.h:
```diff
@@ -0,0 +1,33 @@
+
+/** @} */
+
+#endif /* ZEPHYR_INCLUDE_ARCH_ARM64_ARM_MEM_H_ */
```
ARM MMU commit are addressed in #96575.
I add them here just for compilation. they should be removed with a rebase when #96575 will be merged

erwango commented on drivers/ethernet/phy/phy_microchip_lan8742.c:
```diff
@@ -385,7 +385,7 @@ static int phy_lan8742_init(const struct device *dev)
 	data->state.is_up = false;
 
-	if (device_is_ready(cfg->mdio) {
+	if (!device_is_ready(cfg->mdio)) {
```
Please fix this in previous commit.

arnopo replied:
oops... done

erwango commented on boards/st/stm32mp135f_dk/Kconfig.defconfig:
```diff
@@ -12,4 +12,7 @@ config GPIO_MCP230XX_INIT_PRIORITY
 endif # GPIO_MCP230XX
 
+configdefault NET_L2_ETHERNET
+	default y if NETWORKING
```
Why not doing as other boards:
```
if NETWORKING

config NET_L2_ETHERNET
	default y

endif # NETWORKING
```

arnopo replied:
This update is related to @maass-hamburg comment:  https://github.com/zephyrproject-rtos/zephyr/pull/96134#discussion_r2440383738 

@erwango , @maass-hamburg : Could you please clarify the good syntax here?

maass-hamburg replied:
coming in https://github.com/zephyrproject-rtos/zephyr/pull/98006

erwango commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -36,6 +37,11 @@ extern const struct device *eth_stm32_phy_dev;
+#define ETH_DMA_REGION  DT_INST_PHANDLE(0, memory_regions)
+#define ETH_DMA_LINKER_REGION_NAME LINKER_DT_NODE_REGION_NAME(ETH_DMA_REGION)
+#define __eth_stm32_desc ALIGN_32BYTES(__attribute__((__section__(ETH_DMA_LINKER_REGION_NAME))))
+#define __eth_stm32_buf  ALIGN_32BYTES(__attribute__((__section__(ETH_DMA_LINKER_REGION_NAME))))
```
Two notes on this, none are blocking:
-Use of `ALIGN_32BYTES`. Is that required here or only for `dma_rx_desc_tab` as done for STM32N6 ? Can't it be replaced by `__aligned(32)` to harmonize code ?
-Use of DT region is interesting and maybe can replace existing churn on STM32H7/H7R/S, so we avoid having multiple implementations of the same thing. @juickar Can you log this last point for DV4.4.0 ?

mathieuchopstm replied:
Use `Z_GENERIC_SECTION`.
```suggestion
#define ETH_DMA_REGION  DT_INST_PHANDLE(0, memory_regions)
#define ETH_SECTION Z_GENERIC_SECTION(LINKER_DT_NODE_REGION_NAME_TOKEN(ETH_DMA_REGION))
#define __eth_stm32_desc ALIGN_32BYTES(ETH_SECTION)
#define __eth_stm32_buf  ALIGN_32BYTES(ETH_SECTION)
```

(also, prefer `__aligned(32)` instead of the STM32-specific `ALIGN_32BYTES` if possible)

mathieuchopstm commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -431,6 +431,18 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
+#endif
+#if DT_INST_CLOCKS_HAS_NAME(0, eth_ker)
+	.pclken_ker = {.bus = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bus),
+		       .enr = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bits)},
```
Nit: I'd put series-specific (i.e., MP13) last
```suggestion
#if DT_INST_CLOCKS_HAS_NAME(0, eth_ker)
	.pclken_ker = {.bus = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bus),
		       .enr = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bits)},
#endif
#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32mp13_ethernet)
	.clockselection = DT_NODE_HAS_PROP(0, st_ext_phyclk) ? HAL_ETH1_REF_CLK_RCC
							     : HAL_ETH1_REF_CLK_RX_CLK_PIN,
```

mathieuchopstm commented on drivers/gpio/Kconfig.mcp23xxx:
What is this change required for? I'm struggling to see the link to PR contents

erwango commented on boards/st/stm32mp135f_dk/Kconfig.defconfig:
```diff
@@ -12,4 +12,7 @@ config GPIO_MCP230XX_INIT_PRIORITY
 endif # GPIO_MCP230XX
 
+config NET_L2_ETHERNET
+	default y if NETWORKING
```
@maass-hamburg Can you explicit the need here? `NET_L2_ETHERNET` doesn't have specific depencies today that would benefit from use of `configdefault`.
Also, if actually needed, I'm fine with use of `configdefault`, but then I'd like to expand systematic use at board level rather than doing it in one board.

arnopo commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -515,7 +560,7 @@ int eth_stm32_hal_init(const struct device *dev)
 
 	/* Tx config init: */
-	memset(&tx_config, 0, sizeof(ETH_TxPacketConfig));
+	memset(&tx_config, 0, sizeof(ETH_TxPacketConfigTypeDef));
```
done

arnopo commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -102,6 +108,15 @@ extern ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB];
+#if DT_INST_CLOCKS_HAS_NAME(0, eth_ker)
+	struct stm32_pclken pclken_ker;
+#endif
+#if defined(CONFIG_SOC_SERIES_STM32MP13X)
```
done

etienne-lms commented on boards/st/stm32mp135f_dk/stm32mp135f_dk.dts:
```diff
@@ -73,6 +73,13 @@
 			   <CSI_IO1 0 &mcp23017 3 0>;
 	};
+
+	eth_ram: sram@2FFE0000 {
```
Usually prefer lowercase in node name units.

arnopo commented on drivers/ethernet/phy/phy_microchip_lan8742.c:
```diff
@@ -0,0 +1,443 @@
+#define PHY_LAN8742_DATA(n)                                                                       \
+	static struct phy_lan8742_dev_data phy_lan8742_dev_data_##n = {                           \
+		.dev = DEVICE_DT_INST_GET(n),                                                     \
+		.cb = NULL,                                                                       \
```
done

arnopo commented on drivers/ethernet/phy/phy_microchip_lan8742.c:
```diff
@@ -0,0 +1,443 @@
+
+#define PHY_LAN8742_DEVICE(n)                                                                     \
+	PHY_LAN8742_CONFIG(n);                                                                    \
+	PHY_LAN8742_DATA(n);                                                                      \
```
done

arnopo commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -379,7 +387,8 @@ int eth_stm32_tx(const struct device *dev, struct net_pkt *pkt)
 	} else {
 		/* We need to release the tx context and its buffers */
-		HAL_ETH_TxFreeCallback((uint32_t *)ctx);
+
```
done

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -608,7 +652,7 @@ int eth_stm32_hal_init(const struct device *dev)
 
 	/* Tx config init: */
-	memset(&tx_config, 0, sizeof(ETH_TxPacketConfig));
+	memset(&tx_config, 0, sizeof(tx_config));
```
can be removed https://github.com/zephyrproject-rtos/zephyr/pull/98218#discussion_r2460806481

arnopo replied:
done

maass-hamburg commented on dts/bindings/ethernet/phy/microchip,lan8742.yaml:
```diff
@@ -0,0 +1,20 @@
+
+  reset-gpios:
+    type: phandle-array
+    description: GPIO connected to MAC PHY reset signal pin. Reset is active low.
```
add
```
  default-speeds:
    default: ["10BASE Half-Duplex", "10BASE Full-Duplex", "100BASE Half-Duplex",
              "100BASE Full-Duplex"]
```

etienne-lms reviewed (commented):
No blocking issues.

arnopo commented on boards/st/stm32mp135f_dk/stm32mp135f_dk.dts:
```diff
@@ -73,6 +73,13 @@
+	eth_ram: sram@2FFE0000 {
+		compatible = "zephyr,memory-region", "mmio-sram";
+		reg = <0x2ffe0000 DT_SIZE_K(16)>;
+		#memory-region-cells = <0>;
```
yes probably. the error message is:

```
devicetree error: <Node /sram@2fffc000 in ./zephyrproject/zephyr/boards/st/stm32mp135f_dk/stm32mp135f_dk.dts:77> lacks #memory-region-cells
```

etienne-lms replied:
Ok, that actually comes from Zephyr **edtlib.py** `_phandle_val_list()`.

etienne-lms commented on drivers/ethernet/phy/phy_microchip_lan8742.c:
```diff
@@ -0,0 +1,443 @@
+	struct phy_lan8742_dev_data *const data = dev->data;
+	struct phy_link_state state;
+
+	if (!data->cb) {
```
Actually, this goes against a directive of the Zephyr coding guidelines: see rule 85 in [main rules](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#id42):
> The controlling expression of an if statement and the
> controlling expression of an iteration-statement shall have
> essentially Boolean type

I've proposed https://github.com/zephyrproject-rtos/zephyr/pull/98920 to clarify.

maass-hamburg replied:
I don't think so, the relevant thing is only the result. And as `!` is the boolean NOT operator, the result is always a boolean, even if the "input" isn't a boolean.

etienne-lms replied:
You mean testing `if (data->cb)` would not comply with guidelines while testing `if (!data->cb)` does comply with?

etienne-lms commented on drivers/ethernet/phy/phy_microchip_lan8742.c:
```diff
@@ -0,0 +1,443 @@
+	int ret;
+
+	/* Issue an optional hard reset */
+	if (cfg->gpio_reset.port) {
```
```c
	if (cfg->gpio_reset.port != NULL) {

etienne-lms commented on boards/st/stm32mp135f_dk/stm32mp135f_dk.dts:
```diff
@@ -250,3 +272,32 @@ csi_interface: &dcmipp {
+
+&mdio {
+	pinctrl-0 = <&eth1_mdio_pa2 &eth1_mdc_pg2>;
+	pinctrl-names = "default";
```
Sorry, should be:
```patch
 &mdio {
 	pinctrl-0 = <&eth1_mdio_pa2 &eth1_mdc_pg2>;
 	pinctrl-names = "default";
+	status = "okay";
 
 	eth_phy: ethernet-phy@0 {
 		compatible = "microchip,lan8742";
 		reg = <0>;
 		reset-gpios = <&mcp23017 9 GPIO_ACTIVE_LOW>;
 	};
-
-	status = "okay";
};

etienne-lms reviewed (commented):
LGTM with 1 non-blocking minor issue.

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -431,6 +431,18 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32mp13_ethernet)
+	.clockselection = DT_NODE_HAS_PROP(0, st_ext_phyclk) ? HAL_ETH1_REF_CLK_RCC
+							     : HAL_ETH1_REF_CLK_RX_CLK_PIN,
```
this is later copied over with
```
heth->Init.ClockSelection = cfg->clockselection;
```
why don't you set it a few lines down directly in `static struct eth_stm32_hal_dev_data eth0_data` after `.MediaInterface = STM32_ETH_PHY_MODE(0),`

arnopo replied:
Yes that seems more optimized i removed `clockselection`  from the `eth_stm32_hal_dev_cfg` struct, thanks!

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -580,6 +594,36 @@ int eth_stm32_hal_init(const struct device *dev)
+		LOG_ERR("Failed to configure mac clock. Error %d", ret);
+		return ret;
+	}
+#endif
```
why don't we move it to the enableing of the other clocks in `eth_initialize`? We already set the ptp clock there too, which also is v2 exclusive

etienne-lms commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -102,6 +108,15 @@ extern ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB];
+#endif
+#if DT_INST_CLOCKS_HAS_NAME(0, eth_ker)
+	struct stm32_pclken pclken_ker;
+#endif
```
Let's address this in another P-R.

etienne-lms replied:
@maass-hamburg, FYI I've created https://github.com/zephyrproject-rtos/zephyr/pull/99331 to address your request.

etienne-lms commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -432,6 +441,15 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
+	.pclken_ker = {.bus = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bus),
+		       .enr = DT_INST_CLOCKS_CELL_BY_NAME(0, eth_ker, bits)},
+#endif
+
```
Remove this empty line?

maass-hamburg reviewed (changes requested):
one thing left

arnopo commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -568,6 +579,9 @@ int eth_stm32_hal_init(const struct device *dev)
 	HAL_StatusTypeDef hal_ret = HAL_OK;
+	__maybe_unused const struct eth_stm32_hal_dev_cfg *cfg = dev->config;
+	__maybe_unused uint8_t *desc_uncached_addr;
+	__maybe_unused int ret;
```
done

arnopo commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -568,6 +579,9 @@ int eth_stm32_hal_init(const struct device *dev)
 	HAL_StatusTypeDef hal_ret = HAL_OK;
+	__maybe_unused const struct eth_stm32_hal_dev_cfg *cfg = dev->config;
+	__maybe_unused uint8_t *desc_uncached_addr;
+	__maybe_unused int ret;
```
done

etienne-lms reviewed (approved):
I'd suggest~~ion~~ that commit message for *"boards: st: stm32mp135f_dk: add ethernet nodes"* explicitly mentions the last 16kB of the SYSRAM are used for ethernet buffers.

Non-blocking comments. 
LGTM.

arnopo commented on soc/st/stm32/stm32mp13x/Kconfig:
```diff
@@ -12,3 +12,4 @@ config SOC_SERIES_STM32MP13X
 	select ARM_ARCH_TIMER
 	select SYS_CLOCK_EXISTS
 	select XIP
+	select KERNEL_DIRECT_MAP
```
Hi @djiatsaf-st 

Thanks for the catch!  https://github.com/zephyrproject-rtos/zephyr/pull/100156 solve the issue? should solve the issue
